lib/OpenQA/WebAPI/Auth/OAuth2.pm

```diff
@@ -24,39 +24,69 @@ sub auth_setup {
+        $self->config->{FIXME_oauth2_nickname_from} = 'login';
+    }
+    elsif ('debian_salsa' eq $provider) {
+        $prov_args->{authorize_url} = 'https://salsa.debian.org/oauth/authorize?response_type=code';
+        $prov_args->{token_url} = 'https://salsa.debian.org/oauth/token';
+        $self->config->{FIXME_oauth2_user_url} = 'https://salsa.debian.org/api/v4/user';
+        $self->config->{FIXME_oauth2_token_scope} = 'read_user';
+        $self->config->{FIXME_oauth2_token_label} = 'Bearer';
+        $self->config->{FIXME_oauth2_nickname_from} = 'username';
+    }
+    elsif ('custom' eq $provider) {
+        $prov_args->{authorize_url} = $self->config->{authorize_url};
+        $prov_args->{token_url} = $self->config->{token_url};
+        $self->config->{FIXME_oauth2_user_url} = $self->config->{user_url};
+        $self->config->{FIXME_oauth2_token_scope} = $self->config->{token_scope};
+        $self->config->{FIXME_oauth2_token_label} = $self->config->{token_label};
+        $self->config->{FIXME_oauth2_nickname_from} = $self->config->{nickname_from};
+    }
+    else {
+        croak "Provider $provider not supported";
+    }
 
     $self->app->plugin(
         OAuth2 => {
-            $provider => {
-                key    => $self->config->{key},
-                secret => $self->config->{secret},
-            }});
+            $provider => $prov_args
+        });
 }
 
 sub auth_login {
     my ($self) = @_;
     croak 'Setup was not called' unless $self->config;
 
     my $get_token_args = {redirect_uri => $self->url_for('login')->userinfo(undef)->to_abs};
-    # Note: user:email is GitHub-specific, email may be empty
-    $get_token_args->{scope} = 'user:email';
+    $get_token_args->{scope} = $self->config->{FIXME_oauth2_token_scope};
     $self->oauth2->get_token_p($self->config->{provider} => $get_token_args)->then(
         sub {
-            return unless my $data = shift;
+            return unless my $data = shift;  # redirect to ID provider
 
-            # Get or update user details via GitHub-specific API
+            # Get or update user details
             my $ua    = Mojo::UserAgent->new;
             my $token = $data->{access_token};
-            my $res   = $ua->get('https://api.github.com/user', {Authorization => "token $token"})->result;
+            my $res   = $ua->get($self->config->{FIXME_oauth2_user_url}, {Authorization => $self->config->{FIXME_oauth2_token_label} . " $token"})->result;
             if (my $err = $res->error) {
                 # Note: Using 403 for consistency
                 return $self->render(text => "$err->{code}: $err->{message}", status => 403);
             }
             my $details = $res->json;
             my $user    = $self->schema->resultset('Users')->create_user(
-                $details->{id},
-                nickname => $details->{login},
+                $details->{id} . "@" . $self->config->{provider},
```

Comment:
I was/am ;-) but I assume you mean there's no significant production instance out there, and I've not heard of any project or team relying on it

Reply:
Still a breaking change though. So should be highlighted in the commit message.

Reply:
Highlighting it is the least we should do.

By the way, an automatic database migration can likely not handle this reliably because so far we lack a way to distinguish providers (e.g. we wouldn't be able to tell OpenID users from GitHub users apart).

This brings me to another point: Likely it would make sense *not* to concatenate the string here but instead add a dedicated provider column to the users table. That would maybe make our live easier in the future. Adding another column is not that much effort and the procedure is described here: https://open.qa/docs/#_how_to_update_the_database_schema

Reply:
Agreed, a provider column would probably be best.

Reply:
Agreed. I suggest we do that as a separate change first so that any changes making use of the column come a bit later and are also revertable keeping the updated database scheme in place

Reply:
That's a good idea. I'll create a PR so @phil-hands doesn't need to bother with the database stuff. He can than rebase this PR on top of that.